server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala

```diff
@@ -0,0 +1,99 @@
+import org.reactivestreams.Publisher
+import reactor.core.publisher.Mono
+import reactor.util.function.{Tuple2, Tuples}
+
+object DeDuplicationBlobStore {
+  val DEFAULT_BUCKET = "defaultBucket"
+  val LAZY_RESOURCE_CLEANUP = false
+  val FILE_THRESHOLD = 10000
+}
+
+class DeDuplicationBlobStore @Inject()(dumbBlobStore: DumbBlobStore,
+                                       @Named("defaultBucket") defaultBucketName: BucketName,
+                                       blobIdFactory: BlobId.Factory) extends BlobStore {
+
+  override def save(bucketName: BucketName, data: Array[Byte], storagePolicy: BlobStore.StoragePolicy): Publisher[BlobId] = {
+    Preconditions.checkNotNull(bucketName)
+    Preconditions.checkNotNull(data)
+
+    val blobId = blobIdFactory.forPayload(data)
+
+    SMono(dumbBlobStore.save(bucketName, blobId, data))
+      .`then`(SMono.just(blobId))
+  }
+
+  override def save(bucketName: BucketName, data: InputStream, storagePolicy: BlobStore.StoragePolicy): Publisher[BlobId] = {
+    Preconditions.checkNotNull(bucketName)
+    Preconditions.checkNotNull(data)
+    val hashingInputStream = new HashingInputStream(Hashing.sha256, data)
+    val sourceSupplier: FileBackedOutputStream => SMono[BlobId] = (fileBackedOutputStream: FileBackedOutputStream) => saveAndGenerateBlobId(bucketName, hashingInputStream, fileBackedOutputStream)
+    Mono.using(() => new FileBackedOutputStream(DeDuplicationBlobStore.FILE_THRESHOLD),
+      sourceSupplier,
+      (fileBackedOutputStream: FileBackedOutputStream) => fileBackedOutputStream.reset(),
+      DeDuplicationBlobStore.LAZY_RESOURCE_CLEANUP)
+  }
+
+  private def saveAndGenerateBlobId(bucketName: BucketName, hashingInputStream: HashingInputStream, fileBackedOutputStream: FileBackedOutputStream): SMono[BlobId] =
+    SMono.fromCallable(() => {
+      IOUtils.copy(hashingInputStream, fileBackedOutputStream)
+      Tuples.of(blobIdFactory.from(hashingInputStream.hash.toString), fileBackedOutputStream.asByteSource)
+    })
+      .flatMap((tuple: Tuple2[BlobId, ByteSource]) =>
+        SMono(dumbBlobStore.save(bucketName, tuple.getT1, tuple.getT2))
+          .`then`(SMono.just(tuple.getT1)))
+
+
+  override def readBytes(bucketName: BucketName, blobId: BlobId): Publisher[Array[Byte]] = {
+    Preconditions.checkNotNull(bucketName)
+
+    dumbBlobStore.readBytes(bucketName, blobId)
+  }
+
+  override def read(bucketName: BucketName, blobId: BlobId): InputStream = {
+    Preconditions.checkNotNull(bucketName)
+
+    dumbBlobStore.read(bucketName, blobId)
+  }
+
+  override def getDefaultBucketName: BucketName = defaultBucketName
+
+  override def deleteBucket(bucketName: BucketName): Publisher[Void] = {
+    dumbBlobStore.deleteBucket(bucketName)
+  }
+
+  override def delete(bucketName: BucketName, blobId: BlobId): Publisher[Void] = {
+    Preconditions.checkNotNull(bucketName)
+    Preconditions.checkNotNull(blobId)
+
+    dumbBlobStore.delete(bucketName, blobId)
```

Comment:
deduplicating should not delete, contract tests should be splited.

Reply:
> deduplicating should not delete, contract tests should be splited.

See https://github.com/linagora/james-project/issues/3600, what you describe is not achievable for now.

Reply:
From the DOD:

> `DeDuplicationBlobStore::delete` is a noop.

I do not get your comment regarding that

Reply:
I should update the DOD of #3591 to remove this mention then, as doing a noop will, so far, break deletedMessageVault behavior.

Reply:
it should happen in one issue anyway

Reply:
and you may add a disabled test ensuring this behavior